### DISCUSSION TITLE
Enable output on SI5332

### DISCUSSION THREAD

**Discussion by sophiekovalevsky** [jankae/LibreVNA#281] [General]
Hi @jankae,

I was wondering how you were able to enable the outputs on SI5332. I started working on this and after creating my own functions (Im doing my own synthetizer evaluation board), I discovered that there are certain registers that are not described in https://www.skyworksinc.com/-/media/Skyworks/SL/documents/public/reference-manuals/si5332-rm.pdf

I realized about this just after giving up and test with the clock builder, just to find out that those registers. As soon as I start to use them, I was able to get the expected output. The model used on my board is slightly different one (SI5332B-D-GM3) but still, overall should be the same as SI5332A-GM1. 

Just inspecting the UI, it seems that, you have taken those hidden registers into consideration:
https://github.com/jankae/LibreVNA/blob/7a9d4ddf7cb0eaee02e1b8f7e3bdc6528e3c5306/Software/PC_Application/Device/RegisterAccess/si5332.cpp#L17

However, there is no way reference to these registers. I just made a quick search on this. I wonder how you come up with that.

**Comment by jankae** [comment]
I have honestly no idea anymore. The experiments with the SI5332 were quite a while ago and I do not remember the details. I can try to find the firmware for my SI5332/AD9913/STW81200 eval board, maybe there was a clue in there (but at first glance, I can not find it anymore, might have to look in my backups).

> (Im doing my own synthetizer evaluation board)

I'd love to hear more about that if you are willing to share!

**Reply by sophiekovalevsky** [reply]
> I have honestly no idea anymore. The experiments with the SI5332 were quite a while ago and I do not remember the details. I can try to find the firmware for my SI5332/AD9913/STW81200 eval board, maybe there was a clue in there (but at first glance, I can not find it anymore, might have to look in my backups).

I went into the firmware if the latest version you used was what it was on the repo but it seems to me that firmware itself was only used to setup the usb communication and off load all the logic from STM32 to the UI itself. I checked this myself and the only implementation for SI5332 are only when reading and writing registers itself.


> I'd love to hear more about that if you are willing to share!

I can share for sure, but I guess this first revision of the board is just has so many erros that I would prefer to release on next version. This project is called Kalahari and its my own version of your synthesizer evaluation board. It has been made for me to understand the nitty gritty on hardware and software of some one the components used on LibreVNA, along with the components you have used to test, but also for me to create a rf board in kicad and have some sort of evaluation board similar in shape to these ones:

![lmx2820evm-box](https://github.com/user-attachments/assets/62004f74-a3d9-4cab-91d3-12087f2d8157)

![lmk04832evm-top](https://github.com/user-attachments/assets/ed0a76b7-4b1c-4aa1-9865-bca22ebc1516)

 Its composed of:

- The DDS, AD9913 same as your version
- The synthetizer used on librevna, MAX2871.
- Also, it has STW81200 which Im assuming you wanted to characterize to be able to see if this synthetizer can be used for next librevna revision? The version is the same since there is no other options for synthetizers on ST.
- The clock generator, SI5332, this one as I mentioned,  the part number is different. Im using the 12 version output.
- STM32L412CBU6, slightly different to the one you are using but same family.

The layout is different to the one you have designed, part of the mental training was to figure it out how to trace out the rf signals, which have definitely errors on my end, but this is really exploratory. This is how far I reached.

![before](https://github.com/user-attachments/assets/5475e9b7-390a-4da8-b64b-ca0abddc69a6)

There are differences between my implementation and yours and its the fact that, it contains a power tree to power up on the same board all the ICs mentioned before. Most of the evaluation boards I have found for synthethizer, decides to take apart the power circuit externally so that only the rf part is contained on the pcb itself, but I wanted to have a single pcb with all I need it except for the external clocks that may drive whatever the ICs.

Also, I wanted to explore the use a level shifter to interface the STM32L4 to the AD9913 and skip the use of voltage divider, however, there are still some issues on my end in trying to communicate with AD9913 using this approach, mostly, I presumed,  due the fact that I did not realize that AD9913 uses SDIO instead of SPI, having me to do some rework on that part of the circuit.

Regarding firmware, I wanted to do something different compared to yours, I wanted give it a try to use libopencm3 and stay away from HAL libraries, even the low level ones, for most of the peripherals Kalahari uses: 

- USB
- UART
- SPI / SDIO
- I2C
- GPIO

Which so far I had tested heavily the I2C bus to work with SI5332, UART to have some sort of logging aside of RTT from SWD, GPIO seems to work fine as well, but what Im not really sure yet is the USB communication which would be the same approach you have used to transfer the data natively through USB. My firmware is made in C rather than using a mixture of C along with C++, thats another difference.

I somehow started communication with STW81200 to the point of reading the who ami register but no more than that. The point is that, to actually drive the STW, you need it to drive it from the clock generator, otherwise, the spi communication did not get any response from the device.

One of the other things I wanted to give a try would be to tune in the loop filter on both MAX and STW synthethizer, and select the filtering part on the output of AD9913. But Im just starting out to understand in most of funcionality, how to control SI5332, how to setup the fractional dividers, etc.

Once I can have a better understand of the whole ICs in this version, I will create another revision with the errors found on this one. The goal I guess would be to release a board and firmware that its somehow stable.

This is a render of how kalahari will look like if it would completely assembled, but as you have seen, I have not even reach the 20% of the completion of the project. 

![layout](https://github.com/user-attachments/assets/94999c8f-7657-48bb-813f-ad9e995c9f25)

**Reply by jankae** [reply]
>  but it seems to me that firmware itself was only used to setup the usb communication and off load all the logic from STM32 to the UI itself

I had a closer look now myself and you are right. The idea was to provide a unified GUI interface to the registers of all the chips on my eval board but there was not really much logic in the code (because I wanted to experiment a lot). That GUI had a save/load feature, so I could rapidly load a complete configuration. I think it is very likely that I had some default configuration I would always load which likely also wrote these undocumented registers.
But yeah, my memory on these experiments is not very good, so I can not help you much here.

> I can share for sure, but I guess this first revision of the board is just has so many erros that I would prefer to release on next version.

I think it looks great already and I am sure you will fix all these errors eventually! Good luck with that project :)

**Reply by sophiekovalevsky** [reply]
> I had a closer look now myself and you are right. The idea was to provide a unified GUI interface to the registers of all the chips on my eval board but there was not really much logic in the code (because I wanted to experiment a lot). That GUI had a save/load feature, so I could rapidly load a complete configuration. I think it is very likely that I had some default configuration I would always load which likely also wrote these undocumented registers.
> But yeah, my memory on these experiments is not very good, so I can not help you much here.

I guess this is fine. During these days I worked on this and it seems that the hidden registers serve the purpose of powering down some components in the IC. I wonder how much this clock generator will consume, but probably its for that reason, so thanks so much in anycase.

> I think it looks great already and I am sure you will fix all these errors eventually! Good luck with that project :)

Thanks Jan.
